### PR TITLE
Add support for MD / HTML in room topics

### PR DESCRIPTION
**Pull request by Johennes**
Requires matrix-org/matrix-js-sdk#2272
For https://github.com/vector-im/element-web/pull/21671
MSC: https://github.com/matrix-org/matrix-spec-proposals/pull/3765

Fixes vector-im/element-web#5180

This allows to enter MD / HTML for room topics in the following places:
- `/topic` command
- Room settings overlay
- Space settings overlay

The HTML rendering is displayed in these places:
- `/topic` command
- Room header
- Space home

Fixes: vector-im/element-web#5180

### Screenshots (after)

Screenshot 2022-04-02 at 21 00 06
Screenshot 2022-04-02 at 20 59 19
Screenshot 2022-04-02 at 20 59 35
Screenshot 2022-04-02 at 20 59 49

### Screenshots (before)

Screenshot 2022-04-02 at 21 03 16

---
Here's what your changelog entry will look like:

## ✨ Features
 * Add support for MD / HTML in room topics ([\#8215](https://github.com/matrix-org/matrix-react-sdk/pull/8215)). Fixes vector-im/element-web#5180. Contributed by @Johennes.

Preview: https://pr8215--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

File: src/settings/Settings.tsx
Comment (turt2live) [inline src/settings/Settings.tsx:327]: this will need adding to labs.md in the element-web layer too

**Review by turt2live — COMMENTED** [review]
generally seems fine - just a few points where we might want to consider a different approach.

File: src/HtmlUtils.tsx
Comment (Johennes) [reply src/HtmlUtils.tsx:620]: Hm, it doesn't look like anything changes in the display when I flip that flag. What I do notice though is that emojis are clipping:

Screenshot 2022-04-04 at 20 24 22

File: src/HtmlUtils.tsx
Comment (Johennes) [reply src/HtmlUtils.tsx:614]: This is maybe a question of whether we want to allow extended topics with _richer_ content below the second line as described [here](https://github.com/matrix-org/matrix-spec-proposals/pull/3765#discussion_r842011219). If that's pushing the limits too far, agree that extracting a custom function would be better.

Comment (Johennes) [reply src/HtmlUtils.tsx:614]: Looking into this some more, preventing the conversion of e.g. `# Heading` to `<h1>Heading</h1>` seems tricky with the CommonMark HTML renderer we're using. The conversion also happens when sending the `m.room.topic` event. This function here just sanitizes the HTML for display and performs a few other tweaks.

I pushed a fix that uses CSS to force headings in the timeline topic to the same size and style as normal text via CSS. In the `/topic` dialog, it seems ok to allow headings & friends given that there is more space available here?

Comment (Johennes) [reply src/HtmlUtils.tsx:614]: Screenshot 2022-04-07 at 22 00 29

File: src/HtmlUtils.tsx
Comment (Johennes) [reply src/HtmlUtils.tsx]: Looks like the cropping is unrelated to this PR and stems from the [line height setting](https://github.com/matrix-org/matrix-react-sdk/blob/develop/res/css/views/rooms/_RoomHeader.scss#L153) on the room header.

Even if it didn't make a change, I've set `disableBigEmoji` to `true` now to be on the safe side.

File: src/HtmlUtils.tsx
Comment (turt2live) [reply src/HtmlUtils.tsx:614]: Err, my assumption was that we'd run the resulting HTML through a sanitizer (which is done internally in `bodyToHtml`) with different params. We'd still let the topic contain the HTML (because we can't stop that), but we'd sanitize it more aggressively on the receiving end.

File: res/css/views/rooms/_RoomHeader.scss
Comment (turt2live) [inline res/css/views/rooms/_RoomHeader.scss:158]: I think we can afford 1.2x sized emoji, assuming they don't throw the room header out of alignment

Comment (Johennes) [reply res/css/views/rooms/_RoomHeader.scss:158]: Unfortunately, the 1.2 factor led to the vertical cropping I had mentioned [here](https://github.com/matrix-org/matrix-react-sdk/pull/8215#discussion_r842025310). Using `inherit` also ensures that emoji in HTML topics have the same size as in plain text topics (where they are not wrapped in `.mx_Emoji`).

File: src/HtmlUtils.tsx
Comment (Johennes) [reply src/HtmlUtils.tsx]: Gotcha. I've now extracted a separate function, stripped down the allowed tags and removed the processing of maths and other unrelated things.

File: res/css/views/rooms/_RoomHeader.scss
Comment (turt2live) [reply res/css/views/rooms/_RoomHeader.scss:158]: can we get that documented in-code? 😇

File: src/HtmlUtils.tsx
Comment (turt2live) [inline src/HtmlUtils.tsx:621]: jsdoc formatting:
```suggestion
/** 
 * Turn a room topic into html
 * @param topic plain text topic
 * @param htmlTopic optional html topic
 * @param ref React ref to attach to any React components returned
 * @return The HTML-ified node.
 */
```

File: src/HtmlUtils.tsx
Comment (turt2live) [inline src/HtmlUtils.tsx:642]:
```suggestion
    let emojiBodyElements: ReturnType<typeof formatEmojis>;
```

this should prevent the cast a couple lines later

File: src/HtmlUtils.tsx
Comment (turt2live) [inline src/HtmlUtils.tsx:640]: anytime we ignore a catch we should have a comment for why. Eg: `// ignore because worst case is we assume it's not HTML`

Comment (turt2live) [reply src/HtmlUtils.tsx:640]: ... though we should probably set `isFormattedTopic = false` explicitly so we render *a* topic.

File: src/SlashCommands.tsx
Comment (turt2live) [inline src/SlashCommands.tsx:483]: we want to support headings and such here though? (I suppose as an MVP it doesn't matter)

**Review by turt2live — COMMENTED** [review]
overall looks good - just the minor nitpicks here + a comment in the CSS please :)

File: res/css/views/rooms/_RoomHeader.scss
Comment (Johennes) [reply res/css/views/rooms/_RoomHeader.scss:160]: Ah, good idea, yes!

File: src/HtmlUtils.tsx
Comment (Johennes) [reply src/HtmlUtils.tsx]: Good catch. This was left empty after removing the unrelated code copied from `bodyToHtml`. I set the is-formatted flag back to `false` in the `catch` now.

File: src/SlashCommands.tsx
Comment (Johennes) [reply src/SlashCommands.tsx]: Oh, I mean, yes!! I didn't realize you only wanted the stricter sanitation in the room header. 🤦‍♂️

I added another parameter to control the level of sanitizing and used it to allow the same tags allowed in messages here.

**Comment by Johennes** [comment]
All comments addressed. There is an open discussion on the MSC about whether or not `m.topic` and `m.message` should be allowed to deviate in future (and if not, that would affect the key in the event content). I'm not sure if that should block this PR though.

**Comment by turt2live** [comment]
something appears to have gone very wrong with your merge - suggest a reset & rebase before review can really be completed.

**Comment by Johennes** [comment]
> something appears to have gone very wrong with your merge - suggest a reset & rebase before review can really be completed.

Yeah, I screwed up after pulling in the suggested and applied changes from GitHub. The final diff was actually correct, just the history messed up. Unfortunately, there were a myriad of commits inbetween so the only sensible way I found to fix this was to reset and commit the diff as a single commit. Hope that's ok.

File: res/css/_common.scss
Comment (turt2live) [inline res/css/_common.scss:306]: I feel like we're copy/pasting these styles from somewhere - can we just add the `mx_Dialog` targets to that place? (or pull the markdown changes to this file?)

**Review by turt2live — COMMENTED** [review]
generally still seems fine - just nervous about landing such a change, still. 

For the CSS: if we're not copy/pasting the styles from somewhere, please put a comment surrounding the whole block to note that the `!important` overrides are actually doing a thing.

**Comment by turt2live** [comment]
also you have CI failures + merge conflicts, sorry. Please do a merge commit if possible to avoid having to have the whole thing re-reviewed :)

File: res/css/_common.scss
Comment (Johennes) [reply res/css/_common.scss:306]: They're copied from [_EventTile.scss](https://github.com/matrix-org/matrix-react-sdk/blob/develop/res/css/views/rooms/_EventTile.scss#L466), yes. There are also bits of `markdown-body` styling in `_GroupLayout.scss` and `_ReplyTile.scss`. I could try and extract the styling from `_EventTile.scss` into `_common.scss`? I'm not sure if it wouldn't be better to keep them separate for the future though.

**Comment by Johennes** [comment]
> also you have CI failures + merge conflicts, sorry. Please do a merge commit if possible to avoid having to have the whole thing re-reviewed :)

Builds are green except for the e2e tests. I'm not sure if those failures are due to my changes though as the log doesn't seem related at all:

```
2022-05-09T18:27:50.219Z  * enterprise_erin logs out:
2022-05-09T18:27:50.219Z    * enterprise_erin navigates to user menu ... done
2022-05-09T18:27:50.269Z    * enterprise_erin clicks the 'Sign Out' button ... done
2022-05-09T18:27:50.323Z  * enterprise_erin waits for redirect to config.json (as external page) ... done
failure:  AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
+ actual - expected

+ '1'
- 'some-new-version'
    at /__w/matrix-react-sdk/matrix-react-sdk/test/end-to-end-tests/lib/src/scenarios/update.js:58:25
    at Generator.next (<anonymous>)
    at fulfilled (/__w/matrix-react-sdk/matrix-react-sdk/test/end-to-end-tests/lib/src/scenarios/update.js:20:58)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5) {
  generatedMessage: true,
  code: 'ERR_ASSERTION',
  actual: '1',
  expected: 'some-new-version',
  operator: 'strictEqual'
}
```

I also don't have rights to re-run builds so I cannot retrigger them.

**Comment by t3chguy** [comment]
@Johennes because it is branch-matching with https://github.com/Johennes/element-web/tree/feature/html-topic

**Comment by Johennes** [comment]
> @Johennes because it is branch-matching with https://github.com/Johennes/element-web/tree/feature/html-topic

Ohh, I didn't realize that. 😮

Thanks for the hint. I updated the element-web branch. Could somebody retrigger the e2e CI run?

File: res/css/_common.scss
Comment (turt2live) [reply res/css/_common.scss:306]: a comment to say that they're copied should be sufficient, I think. We *should* extract them properly, but I've been enough of a blocker on this PR already. 

```suggestion
// Styles copied/inspired by GroupLayout, ReplyTile, and EventTile variants.
.mx_Dialog .markdown-body {
```

Comment (Johennes) [reply res/css/_common.scss:307]: Nah, you haven't been a blocker at all. If anything I should have done things right from the start. 🙈

**Review by turt2live — APPROVED** [review]
:shipit:

**Comment by Johennes** [comment]
@turt2live I'm sorry but I'm gonna have to ask you for a re-review. There's been a non-trivial merge conflict that needed resolving and I also added tests to try and approach the 80% coverage threshold (although that seems to no longer be mandatory judging from the latest SonarCloud analysis).

File: src/SlashCommands.tsx
Comment (turt2live) [inline src/SlashCommands.tsx:480]: Code style changes are inbound for this sort of thing:
```suggestion
            const topic = !!content 
                ? ContentHelpers.parseTopicContent(content)
                : { text: _t('This room has no topic.') };
```

File: test/HtmlUtils-test.tsx
Comment (turt2live) [inline test/HtmlUtils-test.tsx:28]: this appears to be testing that the output comes through as text rather than as HTML?

File: test/HtmlUtils-test.tsx
Comment (turt2live) [inline test/HtmlUtils-test.tsx:38]: The return value should be mocked rather than calling the setter - this will bleed to other tests.

File: test/HtmlUtils-test.tsx
Comment (turt2live) [inline test/HtmlUtils-test.tsx:51]: we should also have tests to ensure an unformatted topic (no html component) doesn't get converted unexpectedly.

File: test/SlashCommands-test.tsx
Comment (turt2live) [inline test/SlashCommands-test.tsx:34]: should we be checking that `command` itself is defined too?

File: test/editor/serialize-test.ts
Comment (turt2live) [inline test/editor/serialize-test.ts:97]: missing a blank line above this:
```suggestion

        it('plaintext remains plaintext even when forcing html', function() {
```

**Review by turt2live — CHANGES_REQUESTED** [review]

File: test/SlashCommands-test.tsx
Comment (Johennes) [reply test/SlashCommands-test.tsx:34]: Looks like it'll always return an object: https://github.com/matrix-org/matrix-react-sdk/blob/develop/src/SlashCommands.tsx#L1351

File: test/HtmlUtils-test.tsx
Comment (Johennes) [reply test/HtmlUtils-test.tsx:38]: Ironically I did copy this from another test, blindly assuming this was ok. 🤦‍♂️

File: test/HtmlUtils-test.tsx
Comment (Johennes) [reply test/HtmlUtils-test.tsx]: Yeah, good call. Changed to verifying the actual HTML.

File: test/SlashCommands-test.tsx
Comment (turt2live) [reply test/SlashCommands-test.tsx:34]: the jsdoc doesn't match on that function - can we fix that while we're here?

File: test/HtmlUtils-test.tsx
Comment (turt2live) [inline test/HtmlUtils-test.tsx:39]: sorry, I meant a test where we send `"<b>pizza</b>"` through as plain text (and `null` html) to ensure it comes out as a literal.

File: test/HtmlUtils-test.tsx
Comment (turt2live) [inline test/HtmlUtils-test.tsx:64]: closing bracket needs to be on a new line if we're going to do it this way, which means trailing comma. Alternatively, move the whole `.toEqual(...)` to a new (indented) line.

**Review by turt2live — CHANGES_REQUESTED** [review]
almost there - apologies for the delay in the review.

please note a reopened conversation on the previous review as well.

File: test/SlashCommands-test.tsx
Comment (Johennes) [reply test/SlashCommands-test.tsx:34]: Oh wow, good catch!

**Review by turt2live — APPROVED** [review]
:shipit:

**Comment by Johennes** [comment]
Thanks @turt2live! 🙌

And PS: I don't have write access so could you please merge this for me?

**Comment by Johennes** [comment]
Is this safe to merge now?

**Comment by turt2live** [comment]
I'm going to land this without working cypress tests as I believe the CI is just horribly confused about branches or context or something. Will monitor develop.

**Pull request merged**
